sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -335,33 +333,53 @@ private[sbtplugin] object ScalaJSPluginInternal {
       scalaJSConfigSettings
   )
 
-  private val scalaJSTestFrameworkSettings = Seq(
+  val scalaJSTestSettings: Seq[Setting[_]] = (
+      scalaJSConfigSettings
+  ) ++ Seq(
+      /* Always default to false for scalaJSUseMainModuleInitializer in testing
+       * configurations, even if it is true in the Global configuration scope.
+       */
+      scalaJSUseMainModuleInitializer := false,
+
+      // Use test module initializer by default.
+      scalaJSUseTestModuleInitializer := true,
+
+      scalaJSModuleInitializers ++= {
+        if (scalaJSUseTestModuleInitializer.value) {
+          if (scalaJSUseMainModuleInitializer.value) {
+            throw new MessageOnlyException("You may only set one of " +
+                "scalaJSUseMainModuleInitializer and scalaJSUseTestModuleInitializer to true")
+          }
+
+          Seq(testAdapterInitializer)
+        } else {
+          Seq.empty
+        }
+      },
+
       loadedTestFrameworks := {
         if (fork.value) {
           throw new MessageOnlyException(
               "`test` tasks in a Scala.js project require " +
               "`fork in Test := false`.")
         }
 
-        val frameworks = testFrameworks.value
-        val env = jsEnv.value
-        val files = jsExecutionFiles.value
-
-        val moduleIdentifier = scalaJSLinkerConfig.value.moduleKind match {
-          case ModuleKind.NoModule =>
-            ModuleIdentifier.NoModule
-          case ModuleKind.CommonJSModule =>
-            ModuleIdentifier.CommonJSModule(scalaJSLinkedFile.value.data.getPath)
+        if (!scalaJSUseTestModuleInitializer.value) {
+          throw new MessageOnlyException(
+              "You may only use test related functionality in a Scala.js " +
+              "project, if scalaJSUseTestModuleInitializer := true")
```

Comment:
Same. I would include `in $configName` in the message.

Reply:
Speaking of which, you could change the `fork in Test` in the other check above to `fork in $configName` as well.